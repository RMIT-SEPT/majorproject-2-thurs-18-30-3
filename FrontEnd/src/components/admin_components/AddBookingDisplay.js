import React, {useState, useEffect} from 'react'
import Tooltip from '@material-ui/core/Tooltip';
import moment from 'moment';

import '../../containers/App.css'

//Displays Services in a list view
function AddBookingDisplay({service, employees, avails, createFunc}) {
  const [vals , setVals] = useState({
      time: "",
      date: "",
      employeeName: ""
  });

  const [availabilities, setAvailabilities] = useState([]);

  useEffect(() => {
    setAvailabilities(avails);
  });

  if(service == null)
  {
    return null;
  }
  
  const CURRENT_DATE = moment().format('DD-MM');
  const NEXT_DATE = moment().add(1,'months').format('DD-MM');

  const handleChange = (e) =>{
    const { name, value } = e.target;
    setVals(vals => ({ ...vals, [name]: value }));
  };

  //Retrieve list of employees and find out if they have availability
  const empOptions = employees.map((employee) => {
    var avail = "NO DATA";
    
    if(availabilities)
    {
      availabilities.forEach(msg => {
        if(msg.username === employee.username)
        {
          avail = msg.availability + "";
        }
      });
    }
    return (
      <Tooltip title={avail} value={avail} key = {employee.id} role='tooltip'><option className="employee-option" value={employee.id}>{employee.firstName} {employee.lastName}</option></Tooltip>
    )
  })

  //Button function for adding new booking slot
  const handleSubmit=(e)=>{
    createFunc(service.name, vals.time, vals.date, vals.employeeName);
  };

  const validate = () => {
    if(vals.time !== "" || vals.date !== "" || vals.employeeName !=="")
    {
      return false;
    }
    return true;
  }

  return (
      <main className="add-booking" role="main">
        <div className = 'add-booking-heading'>
            <h1>Add Slot</h1>
        </div>

        <div className = 'add-booking-body'>
            <span>{service.name}</span>
        </div>
        <hr></hr>
        <form className = 'add-booking-form' onSubmit={handleSubmit}>
          <input type="time" name='time' value={vals.time} onChange={handleChange} role='textbox'/>
          <input type="date" name='date' value={vals.date} min={CURRENT_DATE} max={NEXT_DATE} onChange={handleChange} role='textbox'></input>
          <div><label htmlFor='employees'>Assigned Employee No: </label><span role='note'>{vals.employeeName}</span></div>
            <select name="employeeName" id="employees" onChange={handleChange} role ='list' multiple>
              {empOptions}
            </select>
          <button className="actButton" type="submit" disabled={validate()}>Create</button>
        </form>
      </main>
  )
}

export default AddBookingDisplay;
