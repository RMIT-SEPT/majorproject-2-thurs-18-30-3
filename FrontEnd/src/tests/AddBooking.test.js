import React from 'react'

import AddBookingDisplay from '../components/admin_components/AddBookingDisplay.js'
import {StaticRouter} from 'react-router'
import renderer from 'react-test-renderer'
import {render, screen, fireEvent} from '@testing-library/react'
import '@testing-library/jest-dom'
import '@testing-library/jest-dom/extend-expect'

const addFuncMock = jest.fn();
const dummyService = {id: 1, name: 'test', description: 'this is a test'};
const dummyEmployees = [{"username": "tester","userType": "employee","password": "11111111","confirmPassword": 
    "111111111","firstName": "test","lastName": "test","email": "test@test.wiz","mobileNum": "123456","address": "1 test","id": 1}];
const dummyAvailabilities = [{id:1,username:"tester",availability:"free"}];

const setup = () => {
    return render(
      <StaticRouter>
        <AddBookingDisplay service={dummyService} employees={dummyEmployees} avails={dummyAvailabilities} createFunc={addFuncMock}/>
      </StaticRouter>
    )
}
  
  //snapshot test
  test('Test Page Renders', () => {
    const comp = renderer
      .create(
        <StaticRouter>
            <AddBookingDisplay service={dummyService} employees={dummyEmployees} createFunc={addFuncMock}/>
        </StaticRouter>
      ).toJSON();
    
      expect(comp).toMatchSnapshot()
  })

  test('Test Page Contents', () => {
    const container = setup();
    expect(screen.getByRole('heading')).toHaveTextContent('Add Slot');
    expect(screen.getByRole('main')).toHaveClass('add-booking');
    expect(screen.getAllByRole('textbox')[0]).toHaveAttribute('type','time');
    expect(screen.getAllByRole('textbox')[1]).toHaveAttribute('type','date');
    expect(screen.getByRole('button')).toHaveTextContent('Create');
  })

  //Make sure create function isn't called when data is missing
  test('Test Validation', () => {
    const container = setup();
    const btn = screen.getByRole('button');
    const timeBox = screen.getAllByRole('textbox')[0];
    const dateBox = screen.getAllByRole('textbox')[1];

    fireEvent.click(btn);
    expect(addFuncMock).toBeCalledTimes(0);

    fireEvent.change(timeBox, {target: {value: '1:11'}})
    fireEvent.click(btn);
    expect(addFuncMock).toBeCalledTimes(0);

    fireEvent.change(timeBox, {target: {value: '1:11'}})
    fireEvent.change(dateBox, {target: {value: '12/2/20'}})
    fireEvent.click(btn);
    expect(addFuncMock).toBeCalledTimes(0);
    
  })

  test('Test Availability display', () => {
    const container = setup();
    const empOption = screen.getByRole('tooltip');
    fireEvent.click(empOption);
    expect(screen.getByRole('tooltip')).toHaveAttribute('title','free');
  })

  //Make sure create function is called correctly
  test('Test Confirms', () => {
    const container = setup();
    const btn = screen.getByRole('button');
    const timeBox = screen.getAllByRole('textbox')[0];
    const dateBox = screen.getAllByRole('textbox')[1];
    const empList = screen.getByRole('list');

    fireEvent.change(timeBox, {target: {value: '1:11'}});
    fireEvent.change(dateBox, {target: {value: '12/2/20'}});
    fireEvent.change(empList, {target: {value: 'test'}});

    fireEvent.click(btn);
    expect(addFuncMock).toHaveBeenCalled();
  })