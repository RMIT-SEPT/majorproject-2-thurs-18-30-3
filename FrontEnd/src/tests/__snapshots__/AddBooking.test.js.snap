// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Page Renders 1`] = `
<main
  className="add-booking"
  role="main"
>
  <div
    className="add-booking-heading"
  >
    <h1>
      Add Slot
    </h1>
  </div>
  <div
    className="add-booking-body"
  >
    <span>
      test
    </span>
  </div>
  <hr />
  <form
    className="add-booking-form"
    onSubmit={[Function]}
  >
    <input
      name="time"
      onChange={[Function]}
      role="textbox"
      type="time"
      value=""
    />
    <input
      max="18-11"
      min="18-10"
      name="date"
      onChange={[Function]}
      role="textbox"
      type="date"
      value=""
    />
    <div>
      <label
        htmlFor="employees"
      >
        Assigned Employee No: 
      </label>
      <span
        role="note"
      >
        
      </span>
    </div>
    <select
      id="employees"
      multiple={true}
      name="employeeName"
      onChange={[Function]}
      role="list"
    >
      <option
        aria-describedby={null}
        className="employee-option"
        onBlur={[Function]}
        onFocus={[Function]}
        onMouseLeave={[Function]}
        onMouseOver={[Function]}
        onTouchEnd={[Function]}
        onTouchStart={[Function]}
        role="tooltip"
        title="NO DATA"
        value={1}
      >
        test
         
        test
      </option>
    </select>
    <button
      className="actButton"
      disabled={true}
      type="submit"
    >
      Create
    </button>
  </form>
</main>
`;
